File: app/src/main/java/org/worshipsongs/utils/PermissionUtils.java
package org.worshipsongs.utils;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import org.worshipsongs.CommonConstants;
import org.worshipsongs.fragment.AlertDialogFragment;
import org.worshipsongs.worship.R;

import java.util.ArrayList;
import java.util.List;

import static android.content.ContentValues.TAG;
import static android.support.v4.app.ActivityCompat.shouldShowRequestPermissionRationale;

/**
 * Created by vignesh on 24/07/2017.
 */

public final class PermissionUtils
{
    public static boolean isStoragePermissionGranted(final Activity activity)
    {
        if (Build.VERSION.SDK_INT >= 23) {
            if (ContextCompat.checkSelfPermission(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
                Log.v(TAG, "Permission is granted");
                return true;
            } else {
                if (ActivityCompat.shouldShowRequestPermissionRationale(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
                    Log.i(TAG, "Never ask again");
                    AlertDialog.Builder builder = new AlertDialog.Builder(activity);
                    LayoutInflater inflater = (LayoutInflater) activity.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
                    View titleView = inflater.inflate(R.layout.dialog_custom_title, null);
                    TextView titleTextView = (TextView) titleView.findViewById(R.id.title);
                    titleTextView.setText(R.string.permission);

                    TextView messageTextView = (TextView) titleView.findViewById(R.id.subtitle);
                    messageTextView.setTextColor(activity.getResources().getColor(R.color.black_semi_transparent));
                    messageTextView.setText(R.string.storage_permission_meaage);
                    builder.setCustomTitle(titleView);
                    builder.setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener()
                    {
                        @Override
                        public void onClick(DialogInterface dialog, int which)
                        {
                            dialog.cancel();
                        }
                    });
                    builder.setPositiveButton("Settings", new DialogInterface.OnClickListener()
                    {
                        @Override
                        public void onClick(DialogInterface dialog, int which)
                        {
                            dialog.cancel();
                            Intent myAppSettings = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, Uri.parse("package:" + activity.getPackageName()));
                            myAppSettings.addCategory(Intent.CATEGORY_DEFAULT);
                            myAppSettings.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            activity.startActivity(myAppSettings);
                        }
                    });
                    builder.show();
                } else {
                    Log.v(TAG, "Permission is revoked");
                    ActivityCompat.requestPermissions(activity, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 1);
                }
                return false;
            }
        } else { //permission is automatically granted on sdk<23 upon installation
            Log.v(TAG, "Permission is granted");
            return true;
        }
    }


}

File: app/src/main/java/org/worshipsongs/utils/PermissionUtils.kt
package org.worshipsongs.utils

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.provider.Settings
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView

import org.worshipsongs.CommonConstants
import org.worshipsongs.R

import android.content.ContentValues.TAG
import androidx.appcompat.app.AlertDialog

/**
 * Created by vignesh on 24/07/2017.
 */

object PermissionUtils
{
    fun isStoragePermissionGranted(activity: Activity): Boolean
    {
        if (Build.VERSION.SDK_INT >= 23)
        {
            if (ContextCompat.checkSelfPermission(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED)
            {
                Log.v(TAG, "Permission is granted")
                return true
            } else
            {
                if (ActivityCompat.shouldShowRequestPermissionRationale(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE))
                {
                    Log.i(TAG, "Never ask again")
                    val builder = AlertDialog.Builder(activity)
                    val inflater = activity.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
                    val titleView = inflater.inflate(R.layout.dialog_custom_title, null)
                    val titleTextView = titleView.findViewById<View>(R.id.title) as TextView
                    titleTextView.setText(R.string.permission)

                    val messageTextView = titleView.findViewById<View>(R.id.subtitle) as TextView
                    messageTextView.setTextColor(activity.resources.getColor(R.color.black_semi_transparent))
                    messageTextView.setText(R.string.storage_permission_meaage)
                    builder.setCustomTitle(titleView)
                    builder.setNegativeButton(R.string.cancel) { dialog, which -> dialog.cancel() }
                    builder.setPositiveButton(R.string.settings) { dialog, which ->
                        dialog.cancel()
                        val myAppSettings = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, Uri.parse("package:" + activity.packageName))
                        myAppSettings.addCategory(Intent.CATEGORY_DEFAULT)
                        myAppSettings.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                        activity.startActivity(myAppSettings)
                    }
                    builder.show()
                } else
                {
                    Log.v(TAG, "Permission is revoked")
                    ActivityCompat.requestPermissions(activity, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), CommonConstants.STORAGE_PERMISSION_REQUEST_CODE)
                }
                return false
            }
        } else
        { //permission is automatically granted on sdk<23 upon installation
            Log.v(TAG, "Permission is granted")
            return true
        }
    }


}
